### PR TITLE
[WIP] Implementation Sketch: Standardized facilities for lazy evaluation

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
Candidate answer to https://github.com/kokkos/kokkos/issues/4152.

Right now, all throughout Kokkos, often incorrectly, we implement guards for "only calculate this argument to a function of the tooling subsystem if a callback is set." See [this](https://github.com/kokkos/kokkos/blob/master/core/src/impl/Kokkos_Profiling.hpp#L486-L496) for an example of this done wrong, we currently only check if a profile library is loaded, rather than checking that the callback for begin_parallel_for is set (in practice this hasn't mattered much, though it likely will, soon).

To me, the obvious fix for this is to use something to the effect of a promise/future, and handle argument calculation in exactly one place. I propose [here](https://github.com/kokkos/kokkos/blob/master/core/src/impl/Kokkos_Profiling.cpp#L136) for the location. I implement this by adding something called a LazilyEvaluatable, which when invoked calculates a value based on a lambda. `invoke_kokkosp_callback` now forwards arguments that are raw arguments, but invokes LazilyEvaluatables passed to it. This achieves the goals.

Concerns:

1) Does this add overhead? We'll need to measure. But what's happening here is very cheap, a lambda is just a struct that initializes members based on its capture clause. So in these examples, we're constructing a struct containing a reference to a std::string, that's quite cheap. 
2) Handling cases of multiple callbacks needing the same future value. This actually happens, [here](https://github.com/kokkos/kokkos/compare/develop...DavidPoliakoff:feature/lazily-evaluate?expand=1#diff-706bc3e14f4ac802ad28b5bf2f78b3eeeaaac4762cbe1616e00bbd9b48c23e31R153) and [here](https://github.com/kokkos/kokkos/compare/develop...DavidPoliakoff:feature/lazily-evaluate?expand=1#diff-706bc3e14f4ac802ad28b5bf2f78b3eeeaaac4762cbe1616e00bbd9b48c23e31R162) both evaluate the same stuff. I think we could extend these facilities to support this pretty easily, but for this implementation sketch I'm not doing so
3) "But now I have two instances of beginParallelFor." Yep, and that really sucks. The plan would be to use 4.0 to refactor this stuff so that the futures are used consistently. But until 4.0 we will be adding maintenance overhead

Gains

1) We stop leaving guns to our own heads in terms of software quality. Right now it would be so easy to accidentally evaluate something in an inopportune time, the fact that we've done this as correctly as we have is a bit of a miracle
2) Standardization. This is the biggy for me. Right now every time we look at the problem of not adding overhead to Tools, we re-solve the same problems. I'd like to say "the way you guard yourself from calculating an expensive value is to use something std::future-y."

Anyway, let me know what you think

edit: reduced Godbolt, to show the core of what's happening here: https://godbolt.org/z/PGh4hvbMc

### CODE REVIEW

**Comment by crtrott** [comment]
One thing I see here is that we are mixing stuff: in the fence case you actually handed the actual operation (i.e. the fence) to the tool, while for the parallel stuff you don't. 

And if we don't hand the actual evaluation there I don't see a significant benefit here, in particular if the actual call site is already effectively a single place. For example begin_parallel_for itself is only called in two places and only has one implementation, so adding another layer underneath it feels not like a win to me. 

I think the main reason you started to feel like this is because we kinda have to call the fence and allocate callbacks inside the backends instead of some higher level shared function (like deep_copy or parallel_for). Does that ring true?

I think I want to talk a bit more about this, and maybe look through the code with you, we otherwise get to evaluate the pain points in more detail.

**Comment by crtrott** [comment]
OK: had a discussion with David which clarified my thinking a bit more:

On the plus side there is slightly less typing and redundancy. The if based version in the simplest cases is something like:

```c++
if(foo_callback_set()) {
  auto args = compute_args();
  fooCallbackWrapper(args);
}
foo();
```

While this version is something like:
```c++
auto invocable = std::function<arg_t>{[&]() {
  return compute_args();
} };
fooCallbackWrapper(invocable);
foo();
```

One of the benefits of this is that in cases where you have more than one callback (say profiling, and tuning) which use the same arg you can do:
```c++
auto invocable = std::function<arg_t>{[&]() {
  return compute_args();
} };
fooCallbackWrapper(invocable);
barCallbackWrapper(invocable);
foo();
```

instead of 
```c++
if(foo_callback_set() || bar_callback_set()) {
  auto args = compute_args();
  if(foo_callback_set()) 
    fooCallbackWrapper(args);
  if(bar_callback_set())
    barCallbackWrapper(args);
}
```

Now we haven't done this if stuff very cleanly now, but I could see us doing it with a pretty strict pattern. The slight benefit I see coming from this is that I don't have to learn new concepts of "wait shouldn't be this be executed only conditionally, ah ok the fooCallbackWrapper is doing the conditional" (and in Davids proposal the fooCallbackWrapper is using a genericCallbackWrapper internally on top of that). With the if version I see the logic of what gets invoked when right there at the call site. 

Now having such a concept defined in the sense of "this type of function call always behaves this way" is generally fine, but my fear is that if we do these kind of approaches too much (and we already have a lot of them) someone has to know hundreds of little concepts to understand what is going on. 

Basically the tradeoff as I see it here is between more concepts, or more code with redundant info (like the "foo_callback_set" "fooCallbackWrapper" pair), and we need to think about the accumulative tradeoff in some sense. Each individual choice can go either way, but too much on one side and it becomes spaghetti code, too much on the other side you have to dive through layers upon layers of concepts to figure out whats going on. 

As for this particular one here: I am slightly in favor of a cleaned up "if" style approach, but its slightly and I am happy to go with majority opinion on this.

**Comment by nliber** [comment]
Take a look at `[std::apply](https://en.cppreference.com/w/cpp/utility/apply)`.

Correct me if I am wrong, but that is basically what lazy evaluation is here:  you capture the necessary parameters in a tuple-like data structure, and then apply those parameters to the function later.

Now, `std::apply` is C++17, so we'd need to reimplement it (or a simpler subset, as the standard one covers a lot of corner cases we don't have).  Ideally, we only use our home grown one to do the work under C++14 and `std::apply` in C++17 and later.

All the callbacks are just function pointers, correct?  I just want to make sure that `std::function` doesn't have to do an internal allocation (there is a guarantee in the standard that `std::function` can hold a function pointer w/o performing an allocation).

**Comment by DavidPoliakoff** [comment]
@nliber : this isn't quite apply, as I understand it. Nothing is ever held in a tuple, it's always a variadic argument pack. The difference is that now that pack can contain two categories of args

1) Value arguments (float, string, struct, whatever)
2) LazilyEvaluatable arguments (std::function-y guys which I gave a name to, as documentation).

At the time when we invoke the callback, after checking whether it's loaded, the Value arguments are just used as is, while the LazilyEvaluatable arguments are invoked, and their return value is used.

So it's less `apply` to me, more "invoke, except if an argument is itself a function, use the results of that function." I have a reduced godbolt, if that's easier to look at: https://godbolt.org/z/PGh4hvbMc

**Comment by nliber** [comment]
Thanks for correcting me. :-). I think I got it now...

A few things:

You don't want or need to use `std::function`.  That is a runtime binding mechanism (such as a callback registered at run time), and has overhead.  If the object it is holding is not small it will allocate memory.  A lambda performing a complex calculation might be capturing a bunch of references, making it not small.  Better to use a type dedicated to this (`Lazy` in my code sample below), and such a type is very simple to write.

You want to avoid unnecessary copies, which unfortunately means references everywhere.  This can get tricky with lifetimes, forwarding vs. rvalue references (temporaries) vs. l-value references, type deduction, const-ness, etc.  So I went and implemented it. :-)

```
// Note:  using static_cast<T&&>(t) instead of std::forward<T>(t)
// to avoid template instantiations, as this does a lot of forwarding
//
// Some other casts may not look necessary, but are necessary because
// an rvalue ref becomes an lvalue ref whenever it has a name, so
// we need to turn it back into an unnamed temporary

#include <iostream>
#include <type_traits>
#define TRACE std::cout << __PRETTY_FUNCTION__ << '\n'

template <typename F>
struct LazyF {
    // This allows construction in place
    // Handles copy/move if an object of type F is passed in
    // Handles storing a ref if F is a reference
    template <typename... Ts>
    LazyF(Ts&&... ts) : m_f(static_cast<Ts&&>(ts)...) {
        TRACE;
    }

    auto operator()() const {
        TRACE;
        return m_f();
    }

   private:
    F m_f;
};

template <typename FRef>
struct LazyRef {
    static_assert(std::is_reference<FRef>::value, "FRef must be a reference");

    explicit LazyRef(FRef f) : m_f(static_cast<FRef>(f)) { TRACE; }

    auto operator()() const {
        TRACE;
        return static_cast<FRef>(m_f)();
    }

   private:
    FRef m_f;
};

#ifdef LAZY_CAN_STORE_BY_VALUE
template <typename F>
using Lazy = LazyF<F>;
#else
template <typename FRef>
using Lazy = LazyRef<FRef>;
#endif

#ifdef LAZY_CAN_STORE_BY_VALUE
// Deliberately returns either a Lazy<Fn> (when f is an lvalue ref) or
// a Defer<Fn&> (when f is an rvalue ref), because that is
// how F is deduced
//
// This has the effect of storing f if an rvalue ref was passed in,
// or storing a reference to f if an lvalue ref was passed in.
template <typename F>
auto make_lazy(F&& f) {
    TRACE;
    return Lazy<F>(static_cast<F&&>(f));
}
#else
template <typename F>
auto make_lazy(F& f) {
    TRACE;

    // The cast, while not strictly necessary,
    // is there to match make_lazy(F&&)
    return Lazy<F&>(static_cast<F&>(f));
}

template <typename F>
auto make_lazy(F&& f) {
    TRACE;

    // The cast is necessary because F is a
    // forwarding reference
    return Lazy<F&&>(static_cast<F&&>(f));
}
#endif

// This matches anything for which we don't have an exact overload
template <typename T>
T&& extract(T&& t) {
    TRACE;
    return static_cast<T&&>(t);
}

// Because extract(T&&) is match for anything that is exact,
// we need an overload for const lvalue ref, non-const lvalue ref
// and for (non-const) rvalue ref of Lazy

#ifdef LAZY_CAN_STORE_BY_VALUE
template <typename F>
auto extract(Lazy<F> const& d) {
    TRACE;
    return d();
}

template <typename F>
auto extract(Lazy<F>& d) {
    TRACE;
    return d();
}

#else
template <typename F>
auto extract(Lazy<F> const&) = delete;

template <typename F>
auto extract(Lazy<F>&) = delete;
#endif

template <typename F>
auto extract(Lazy<F>&& d) {
    TRACE;
    return d();
}

extern bool should_invoke;
bool should_invoke = true;

template <typename F, typename... Ts>
void defer_invoke(F&& f, Ts&&... ts) {
    TRACE;

    // the static_cast<void> is only because I'm getting
    // paranoid about [[nodiscard]]
    if (should_invoke)
        static_cast<void>(
            static_cast<F&&>(f)(extract(static_cast<Ts&&>(ts))...));
}

inline void takes_two_ints(int, int) { TRACE; }

inline void test_defer() {
    TRACE;

    std::cout << "\nnum, lazy\n";
    defer_invoke(takes_two_ints, 2, make_lazy([] { return 3; }));

    std::cout << "\nlazy, num\n";
    defer_invoke(takes_two_ints, make_lazy([] { return 5; }), 7);

    std::cout << "\nlazy, lazy\n";
    defer_invoke(takes_two_ints, make_lazy([] { return 11; }),
                 make_lazy([] { return 13; }));

    std::cout << "\nnum, num\n";
    defer_invoke(takes_two_ints, 17, 19);

    std::cout << "\nnon-const lambda\n";
    auto l = [] { return 23; };
    make_lazy(l);

    std::cout << "\nconst lambda\n";
    auto const cl = [] { return 29; };
    make_lazy(cl);

    std::cout << "\nfunction ptr\n";
    make_lazy(+cl);

    std::cout << "\nlvalue function ptr\n";
    make_lazy(takes_two_ints);
}

```

If `LAZY_CAN_STORE_BY_VALUE` is defined, then temporary lambdas are stored by value in `LazyF`.  I believe that version gets all the lifetime issues correct.  TODO: make `LazyF` copy construction work correctly (or just make it a non-moveable non-copyable type), as there is a weird case because `LazyF(LazyF&)` calls the templated constructor, not the copy constructor.

If `LAZY_CAN_STORE_BY_VALUE` is not defined, then temporary lambdas/structs are stored by reference in `LazyRef`.  I still need to do some more tests to make sure lifetime issues are correct; i.e., do the all the temporaries inside the `defer_invoke` call actually have the correct lifetimes (core keeps changing the lifetime extension rules...).

If you have any questions, don't hesitate.  While the code may look simple and straightforward, it is written that way because I've internalized a lot of the C++ rules and gotchas over the years and I might be writing it with implicit knowledge that really should be explicit because it is tricky (or I might just have gotten it wrong).

**Comment by nliber** [comment]
I left TRACE in as it is very helpful to see what the exact types are with regards to references and constness.  I found that gcc output is less verbose and hence more readable than clang output, unless you care exactly what like a lambda is defined on.

**Comment by nliber** [comment]
I now believe not having `LAZY_CAN_STORE_BY_VALUE` defined works correctly with respect to lifetimes.

**Comment by nliber** [comment]
I updated my implementation:

```
// This is designed so that all the parameters in one-liner calls to
// defer_invoke(...) have the correct lifetime (all remain alive until the
// end of the call) without doing any unnecessary copies or moves (basically
// uses references everywhere it can).
//
// TRACE should be removed entirely before putting this into production!
//
// For forwarding references:
//  template <typename T> void forwarding_reference(T&& t) { /* ... */ }
//      • If t is an l-value reference to U, T deduces to U&
//      • If t is an r-value reference to U, T deduces to U, not U&&
//      • static_cast<T&&>(t) is the same as std::forward<T>(t) (with fewer
//        template instantiations), and turns t into either a U& or U&&
//      • If you are passed a forwarding reference, you should forward it
//        before using it
//
// • auto always denotes a value
// • decltype(auto) always denotes a reference (either U& or U&&)

#include <iostream>
#include <type_traits>
#define TRACE std::cout << __PRETTY_FUNCTION__ << '\n'

template <typename F>
struct Lazy {
    static_assert(std::is_reference<F>::value, "F must be a reference type.");

    // F is a reference type (either Fn& or Fn&&), so the cast is doing a
    // std::forward<F>(f)
    explicit Lazy(F f) : m_f{static_cast<F>(f)} { TRACE; }

    // The operator itself is r-value reference qualified as we only want to
    // work on temporary Lazy instances
    decltype(auto) operator()() const&& {
        TRACE;
        return static_cast<F>(m_f)();
    }

   private:
    F m_f;
};

// F&& deduces as a forwarding reference
// So we need the F& as well to correctly
// handle an l-value reference
//
// This is needed for both deduction guides and make_lazy
#if __cpp_deduction_guides
template <typename F>
explicit Lazy(F&)->Lazy<F&>;

template <typename F>
explicit Lazy(F &&)->Lazy<F&&>;
#endif

template <typename F>
Lazy<F&> make_lazy(F& f) {
    TRACE;
    return Lazy<F&>(f);
}

template <typename F>
Lazy<F&&> make_lazy(F&& f) {
    TRACE;
    return Lazy<F&&>(static_cast<F&&>(f));
}

// This matches anything for which we don't have an exact overload
template <typename T>
T&& extract(T&& t) {
    TRACE;
    return static_cast<T&&>(t);
}

// We only want to work on temporary Lazy<F> instances,
// so (ignoring volatile) we delete the two that take it by l-value reference
template <typename F>
decltype(auto) extract(Lazy<F> const&) = delete;

template <typename F>
decltype(auto) extract(Lazy<F>&) = delete;

template <typename F>
decltype(auto) extract(Lazy<F>&& d) {
    TRACE;
    return static_cast<Lazy<F>&&>(d)();
}

extern bool should_invoke;
bool should_invoke = true;

template <typename F, typename... Ts>
void defer_invoke(F&& f, Ts&&... ts) {
    TRACE;

    // the static_cast<void> is only because I'm getting
    // paranoid about [[nodiscard]]
    if (should_invoke)
        static_cast<void>(
            static_cast<F&&>(f)(extract(static_cast<Ts&&>(ts))...));
}

```

**Comment by DavidPoliakoff** [comment]
Do we want to standardize our lazy evaluation mechanism
SF F N A SA
1 2 3 1 0

Not strong enough to implement
